BungeeCord-Patches/0054-Speed-up-packet-construction.patch

```diff
@@ -0,0 +1,337 @@
+                     map( ProtocolConstants.MINECRAFT_1_8, 0x03 )
+             );
+             TO_CLIENT.registerPacket(
+                     LoginPayloadRequest.class,
++                    LoginPayloadRequest::new, // Waterfall - speed up packet construction
+                     map( ProtocolConstants.MINECRAFT_1_13, 0x04 )
+             );
+ 
+             TO_SERVER.registerPacket(
+                     LoginRequest.class,
++                    LoginRequest::new, // Waterfall - speed up packet construction
+                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
+             );
+             TO_SERVER.registerPacket(
+                     EncryptionResponse.class,
++                    EncryptionResponse::new, // Waterfall - speed up packet construction
+                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
+             );
+             TO_SERVER.registerPacket(
+                     LoginPayloadResponse.class,
++                    LoginPayloadResponse::new, // Waterfall - speed up packet construction
+                     map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
+             );
+         }
+@@ -344,7 +381,7 @@ public enum Protocol
+ 
+         private final int protocolVersion;
+         private final TObjectIntMap<Class<? extends DefinedPacket>> packetMap = new TObjectIntHashMap<>( MAX_PACKET_ID );
+-        private final Constructor<? extends DefinedPacket>[] packetConstructors = new Constructor[ MAX_PACKET_ID ];
++        private final java.util.function.Supplier<? extends DefinedPacket>[] packetConstructors = new java.util.function.Supplier[ MAX_PACKET_ID ]; // Waterfall - speed up packet construction
+     }
+ 
+     @Data
+@@ -412,21 +449,21 @@ public enum Protocol
+                 throw new BadPacketException( "Packet with id " + id + " outside of range " );
+             }
+ 
+-            Constructor<? extends DefinedPacket> constructor = protocolData.packetConstructors[id];
++            java.util.function.Supplier<? extends DefinedPacket> constructor = protocolData.packetConstructors[id]; // Waterfall - speed up packet construction
+             try
+             {
+-                return ( constructor == null ) ? null : constructor.newInstance();
+-            } catch ( ReflectiveOperationException ex )
++                return ( constructor == null ) ? null : constructor.get(); // Waterfall - speed up packet construction
++            } catch ( Exception ex ) // Waterfall - speed up packet construction
+             {
+                 throw new BadPacketException( "Could not construct packet with id " + id, ex );
+             }
+         }
+ 
+-        private void registerPacket(Class<? extends DefinedPacket> packetClass, ProtocolMapping... mappings)
++        public <P extends DefinedPacket> void registerPacket(Class<P> packetClass, java.util.function.Supplier<P> constructor, ProtocolMapping... mappings) // Waterfall - speed up packet construction
```

Comment:
private -> public is a leftover from my fork, I didn't intend to change access modifiers.